### DISCUSSION TITLE
SVG animation (features?)

### DISCUSSION THREAD

**Discussion by pgilfernandez** [friction2d/friction#293] [Q&A]
I've testing the animated SVG format, it can do a lot of things, great for Friction!

Anyway, I would like to know what this format features has implemented and which not yet.
For instance I found out that it can't export mask shapes using DstIn... is there any working way to export masks to animated SVG?

Thanks!

**Comment by rodlie** [comment]
> Anyway, I would like to know what this format features has implemented and which not yet.

The exporter should warn if you use features not supported.

I don't have a list of features supported, will need to check. In general all the basic stuff works.

> For instance I found out that it can't export mask shapes using DstIn... is there any working way to export masks to animated SVG?

Masking was introduced in 0.9.6.1 (backported from v1), it works fine, used it at work several times today.

https://friction.graphics/releases/friction-0961.html#svg-changes

I could probably make a tutorial.

**Reply by rodlie** [reply]
btw, remember to promote group(s) with masks to layer(s), and that you use white as color.

**Reply by pgilfernandez** [reply]
Working like a charm!
I needed to group the mask and the masked shapes into a group and then promote it to layer 👏🏻

**Reply by pgilfernandez** [reply]
And what about "shape following a path"?
The SVG standard allows it but I could not do it with Friction... any trick to do it or it can't just do it yet?

Thanks?

**Reply by rodlie** [reply]
See https://github.com/friction2d/friction/issues/173

**Reply by rodlie** [reply]
Should be able to just merge the `svg-follow-path` branch into what branch you use, been a while since I looked at it. Worst case I could update the branch it anyone want to test.

https://github.com/friction2d/friction/compare/main...svg-follow-path

**Reply by pgilfernandez** [reply]
Great! Tomorrow I'll try and let you know.
I've been trying today SVG export of common and advanced features and it can do most of them, follow path was missing but I see you've done lot of work on it. I'll try it out tomorrow =)

**Reply by pgilfernandez** [reply]
> Should be able to just merge the `svg-follow-path` branch into what branch you use, been a while since I looked at it. Worst case I could update the branch it anyone want to test.
> 
> [main...svg-follow-path](https://github.com/friction2d/friction/compare/main...svg-follow-path)

I have just tested, it worked fine with some issues that I guess you are aware of:
- I created some keyframes for the "complete" parameter (0.0 at 0f, 1.0 at 60f and back to 0.0 at 120f), the exported SVG just animates from 0.0 to 1.0 but not from 1.0 to 0.0
- as you commented before, rotation didn't work

BTW, in case you decide to improve it let me know and I'll be glad to test.
Here is my test just in case you need it: [svga_follow_path.zip](https://github.com/user-attachments/files/17610942/svga_follow_path.zip)

**Reply by rodlie** [reply]
The rotate stuff is currently a blocker (and there are probably other issues), if not this would be part of v1.

It's currently on the v1.1 roadmap, but will probably be delayed even further, as the v1.1 roadmap is already not realistic.

**Reply by pgilfernandez** [reply]
IMHO, I would include it in v1 as the basic feature works (well, maybe you would need to polish why the keyframes are not working perfectly as I showed you in my previous test), I mean, it works pretty good and it is always better to have a follow path without rotation than nothing... Bear in mind that lot of times the rotation is not needed but, well, it's up to you, of course.

**Reply by pgilfernandez** [reply]
Here you have a brief study of the "svg follow path rotation" problem... I hope it helps you =)

https://github.com/orgs/friction2d/discussions/326

**Reply by pgilfernandez** [reply]
> Masking was introduced in 0.9.6.1 (backported from v1), it works fine, used it at work several times today.

I was using masking today and found out that if you demote the layer to group it not only doesn't export the mask but it doesn't behave the same with layers below and out of the group/layer you "have masked".

But, well, apart from that I was thinking about the "obligation" to use white color... I was testing other colors and it seems what it translates to opacity is the "saturated color", that is, the grey value your color has. But if you use the alpha channel it lowers too the opacity...
What I mean and would like to suggest is if it would be a good idea to override RGB color with white at SVG export time so that any color you pick it behaves as normal Friction does and let the mask opacity be handled just by the alpha color channel, once again as normal Friction does.

Summing it up, **why don't override any color you have set to a `DstIn` layer when exporting to SVG turning it into white,** that is, the color you need for the mask to work?
It would help users to match normal Friction behavior

**Reply by rodlie** [reply]
> I was using masking today and found out that if you demote the layer to group it not only doesn't export the mask but it doesn't behave the same with layers below and out of the group/layer you "have masked".

Masking is a compositing operation and is reserved for layers.

> why don't override any color you have set to a DstIn layer when exporting to SVG turning it into white, that is, the color you need for the mask to work?

Not sure about that one, first we can't really assume everything in a DstIn layer should be white, should we change all fill to white? is that also true for stroke? we could also have several groups/layers (with misc objects) under a main DstIn layer.

**Reply by pgilfernandez** [reply]
> Masking is a compositing operation and is reserved for layers.

Do you mean also for Friction or are you talking about SVG export?

> Not sure about that one, first we can't really assume everything in a DstIn layer should be white, should we change all fill to white? is that also true for stroke? we could also have several groups/layers (with misc objects) under a main DstIn layer.

Umm, let me test all possibilities in depth tomorrow and I'll let you know.

**Reply by pgilfernandez** [reply]
Short answer: I found so many different behaviors while testing and mixing the following parameters:
- color channels (RGB)
- alpha one
- fill
- stroke
- groups
- layers
- group/layer ordering

... so, I'll need more time to explore and document them all to have an overall view of what's happening but the truth to be told is that I find different behaviors between how normal Friction renders and how SVG export does so it's better to wait a little bit more and see "my results"...

**Comment by Dinesh0N** [comment]
Inkscape animation plugin for reference:
https://imvenx.github.io/inkaweb/#/

**Reply by pgilfernandez** [reply]
> Inkscape animation plugin for reference: https://imvenx.github.io/inkaweb/#/

Thanks!
I was using https://lottiefiles.com/supported-features as it is a simple feature table and a way to side by side compare Lottie with animated SVGs
